### ISSUE TITLE
date_histogram issue when using "pre_zone_adjust_large_interval" and a timezone with DST

### ISSUE THREAD

**Issue by orenash** [elasticsearch/elasticsearch#9491]
Hey there.

Since we upgraded to version 1.3.7, we noticed that we sometimes get weird double buckets when running our date_histogram aggregations.

After trying some things we figured out it only happens when we set `pre_zone_adjust_large_interval` to true, `pre_zone` to our local time zone (which uses DST), and `interval` is 'day' or bigger.

In the aggregration's result we see that we get two buckets corresponding to the same interval, but with an hour difference between their keys.

Here is a simple example that should reproduce this issue:

```javascript
POST /test/t
{
    "d": "2014-10-08T13:00:00Z"
}

POST /test/t
{
    "d": "2014-11-08T13:00:00Z"
}

GET /test/_search?size=0
{
   "aggs": {
      "test": {
         "date_histogram": {
            "field": "d",
            "interval": "year",
            "pre_zone": "Asia/Jerusalem",
            "pre_zone_adjust_large_interval": true
         }
      }
   }
}
```

And the result:

```javascript
"aggregations": {
      "test": {
         "buckets": [
            {
               "key_as_string": "2013-12-31T21:00:00.000Z",
               "key": 1388523600000,
               "doc_count": 1
            },
            {
               "key_as_string": "2013-12-31T22:00:00.000Z",
               "key": 1388527200000,
               "doc_count": 1
            }
         ]
      }
   }
```

As you can see, although the two timestamps unarguably belong to the same year, they're put into different buckets. As we figured out, it happens because the timezone offset is different for the two timestamps as one of them occurs when DST is on and the other is not.

We use `Asia/Jerusalem` timezone but the same problem happens in every timezone with DST (eg CET), as long as `pre_zone_adjust_large_interval` is used (When it's not used both documents will go to the same bucket `"2014-01-01T00:00:00.000Z"`).

This problem never happened before we upgraded to 1.3.7. Moreover, while digging into the code we figured out that this bug is a direct result of the proposed fix for issue #8339. While that fix indeed solved the timezone problem in 'hour' intervals around DST switch, it caused the new problem with bigger intervals.

The problem is also in version 1.4.2.

**Comment by orenash** [comment]
I had the time to look into this issue and figure out the exact reasons for this to happen and how it might be fixed.

This issue is super confusing and frustrating, so get ready for lots of fun... :)

When a key is to be rounded for a date histogram, the latest code in `TimeZoneRounding` works in three different strategies:
1. `UTCTimeZoneRoundingFloor` (when `pre_zone` is UTC) - the key is rounded directly, no problems here.
2. `TimeTimeZoneRoundingFloor` (when `pre_zone` is not UTC, `interval<=hour` or `pre_zone_adjust_large_interval=true`) - the timezone offset is added to the key before rounding it, and then subtracted from the rounded key in order to put it back in the correct timezone.
3. `DayTimeZoneRoundingFloor` (when `pre_zone` is not UTC, `interval>=day`, `pre_zone_adjust_large_interval=false`) - the timezone offset is added before rounding and the rounded value remains in UTC. No problems here, but we find it cleaner and easier in our application to get the buckets marked with timestamp in our local timezone, regardless of the histogram's interval, that's why we want to keep using `pre_zone_adjust_large_interval`.

Notice that only `DayTimeZoneRoundingFloor` has no problems, that's why this issue and #8339 have no concern to users who don't use `pre_zone_adjust_large_interval` with large intervals.

Two similar problems may arise in `TimeTimeZoneRoundingFloor`, when subtracting the `pre_zone` offset from the rounded key:
1. If adding the timezone offset to the key results in a timestamp with different offset. This is relevant only for timestamps around the DST switch. For example `2014-10-25T22:00:00Z` has +3 offset in `Asia/Jerusalem` and after adding 3 hours to it we get a timestamp with +2 offset.
2. If *rounding* the key results in a timestamp with different offset. This is relevant only for intervals larger than hour (assuming DST switch occurs on round hours only). For example `2014-08-08T00:00:00Z` has +3 offset in `Asia/Jerusalem` and after rounding to year interval we get +2 offset.

Issue #8339 addressed the first problem. The fix (#8655) modified `TimeTimeZoneRoundingFloor` so the subtracted offset from the rounded key is the same offset that was originally added to the key (instead of the offset at the time of the rounded key itself), and thus solving the first problem.

But what about the second problem with large intervals?
Prior to #8655, the subtracted offset was the offset at the time of the rounded key, so all keys that resulted with the same rounded key went to the same bucket, as expected. In rare cases the bucket was marked with an incorrect timestamp (only on `interval=day` and on the day of the DST switch), but this was a minor unnoticed problem.
After #8655 the problem has a bigger effect: if two original keys are rounded to the same key, but originally had different offsets (eg one is before DST switch and the other is after it), they are sent into different buckets.

So how to solve this?
When `interval>=day`, the rounded key will always point on a UTC midnight. We expect the subtraction to correct it so it will point to the same midnight, but in `pre_zone`.
Notice that in all cases, the correct offset to be subtracted from the rounded key is the offset of `pre_zone` at the desired timestamp, which we are only trying to calculate now. (confusing, right?). For example, in order to get from `2014-11-11T00:00:00Z` to `2014-11-10T22:00:00Z` (which is midnight in Jerusalem), the correct offset to subtract is the offset on `2014-11-10T22:00:00Z`, but we don't know it yet...
After trying some things, I figured out the correct way to get this offset is to use the [getOffsetFromLocal](http://joda-time.sourceforge.net/apidocs/org/joda/time/DateTimeZone.html#getOffsetFromLocal(long)) method of the timezone class and pass it the rounded key.

I changed the code to use this method and all the tests pass, including tests I added for the issue that fail on the latest code.

But! (of course there is a but), I found another rare case which isn't covered by the current tests and fails when using `getOffsetFromLocal` (and succeeds with latest code). This is the case immediately after switching DST on->off, the two hours before and after the switch are "ambiguous". If interval>=day, this is irrelevant because rounding takes us to midnight anyway. Nevertheless, if interval=hour, there is no way for `getOffsetFromLocal` to distinguish between them so this means they both go into the same bucket. That's probably OK for most applications, because hour histogram around DST switch is confusing to display whatsoever. But since I am pedantic, I want to solve this case as well... ^_^

So my solution is to split `TimeTimeZoneRoundingFloor` into two strategies, one for small intervals, which should work as in the latest code, and one for large intervals, which uses `getOffsetFromLocal`.

I will post a pull request with my fix and my additional tests as soon as I figure out if anything should also be changed in the strategies that deal with intervals such as "2d" (I think this problem is irrelevant there). Also if anyone has a cleaner solution (e.g. doesn't split the strategy) please let me know.

Here are the tests I added in order to cover `pre_zone_adjust_large_interval` functionality and the double buckets bug:

```java
    @Test
    public void testAdjustPreTimeZone() {
        Rounding tzRounding;

        // Day interval
        tzRounding = TimeZoneRounding.builder(DateTimeUnit.DAY_OF_MONTH).preZone(DateTimeZone.forID("Asia/Jerusalem")).preZoneAdjustLargeInterval(true).build();
        assertThat(tzRounding.round(time("2014-11-11T17:00:00", DateTimeZone.forID("Asia/Jerusalem"))), equalTo(time("2014-11-11T00:00:00", DateTimeZone.forID("Asia/Jerusalem"))));
        // DST on
        assertThat(tzRounding.round(time("2014-08-11T17:00:00", DateTimeZone.forID("Asia/Jerusalem"))), equalTo(time("2014-08-11T00:00:00", DateTimeZone.forID("Asia/Jerusalem"))));
        // Day of switching DST on -> off
        assertThat(tzRounding.round(time("2014-10-26T17:00:00", DateTimeZone.forID("Asia/Jerusalem"))), equalTo(time("2014-10-26T00:00:00", DateTimeZone.forID("Asia/Jerusalem"))));
        // Day of switching DST off -> on
        assertThat(tzRounding.round(time("2015-03-27T17:00:00", DateTimeZone.forID("Asia/Jerusalem"))), equalTo(time("2015-03-27T00:00:00", DateTimeZone.forID("Asia/Jerusalem"))));

        // Month interval
        tzRounding = TimeZoneRounding.builder(DateTimeUnit.MONTH_OF_YEAR).preZone(DateTimeZone.forID("Asia/Jerusalem")).preZoneAdjustLargeInterval(true).build();
        assertThat(tzRounding.round(time("2014-11-11T17:00:00", DateTimeZone.forID("Asia/Jerusalem"))), equalTo(time("2014-11-01T00:00:00", DateTimeZone.forID("Asia/Jerusalem"))));
        // DST on
        assertThat(tzRounding.round(time("2014-10-10T17:00:00", DateTimeZone.forID("Asia/Jerusalem"))), equalTo(time("2014-10-01T00:00:00", DateTimeZone.forID("Asia/Jerusalem"))));

        // Year interval
        tzRounding = TimeZoneRounding.builder(DateTimeUnit.YEAR_OF_CENTURY).preZone(DateTimeZone.forID("Asia/Jerusalem")).preZoneAdjustLargeInterval(true).build();
        assertThat(tzRounding.round(time("2014-11-11T17:00:00", DateTimeZone.forID("Asia/Jerusalem"))), equalTo(time("2014-01-01T00:00:00", DateTimeZone.forID("Asia/Jerusalem"))));

        // Two timestamps in same year ("Double buckets" bug in 1.3.7)
        tzRounding = TimeZoneRounding.builder(DateTimeUnit.YEAR_OF_CENTURY).preZone(DateTimeZone.forID("Asia/Jerusalem")).preZoneAdjustLargeInterval(true).build();
        assertThat(tzRounding.round(time("2014-11-11T17:00:00", DateTimeZone.forID("Asia/Jerusalem"))),
                equalTo(tzRounding.round(time("2014-08-11T17:00:00", DateTimeZone.forID("Asia/Jerusalem")))));
    }
```


Here is a test that covers the ambiguous hours bug:
```java
    @Test
    public void testAmbiguousHoursAfterDSTSwitch() {
        Rounding tzRounding;

        tzRounding = TimeZoneRounding.builder(DateTimeUnit.HOUR_OF_DAY).preZone(DateTimeZone.forID("Asia/Jerusalem")).preZoneAdjustLargeInterval(true).build();
        assertThat(tzRounding.round(time("2014-10-25T22:30:00", DateTimeZone.UTC)), equalTo(time("2014-10-25T22:00:00", DateTimeZone.UTC)));
        assertThat(tzRounding.round(time("2014-10-25T23:30:00", DateTimeZone.UTC)), equalTo(time("2014-10-25T23:00:00", DateTimeZone.UTC)));
    }
```

**Comment by synhershko** [comment]
I can confirm I've seen this issue before - back in the facet days with 0.90. I recall opening an issue on this but can't find it now. Good catch!

**Comment by cbuescher** [comment]
Thanks for the great test case. Since I'm currently trying to clean up the time zone management for 2.0 I included your test case from the first comment. I can reproduce the issue on 1.4 and current 2.0 branch and I was able to make it pass using the following changes to TimeZoneRounding.TimeTimeZoneRoundingFloor:

'''
public long roundKey(long utcMillis) {
            long local = preTz.convertUTCToLocal(utcMillis);
            return preTz.convertLocalToUTC(field.roundFloor(local), true, utcMillis);
}
'''

This basically follows your suggestions and the work in #9637. I will also add your other Rounding tests there and see if they pass with my intended implementation. However, the `pre_zone_adjust_large_interval` will hopefully be gone for 2.0.

**Comment by orenash** [comment]
Yes, the code for roundKey that you included looks exactly like my final solution, that's great.
Can this fix also be included in 1.x? (As I said I can post a PR for this myself, but if you plan to include it in 2.0 I guess it's better you manage it).

About removing `pre_zone_adjust_large_interval`, from what I understood, you plan to always return times in UTC on buckets. Which is equivalent to setting `pre_zone_adjust_large_interval=true` now. Am I right? If so, it means you can use the tests I provided for `pre_zone_adjust_large_interval=true` to test 2.0 (instead of the existing test that assume `pre_zone_adjust_large_interval=false` and will fail on your new code).

**Comment by cbuescher** [comment]
> Which is equivalent to setting pre_zone_adjust_large_interval=true

Yes, I think that is what it comes down to. I was able to use your test cases above with minor modifications to test on 2.0 branch. I think the plan is to backport the parts that are bug fixes to 1.x branch, but you can also propose a PR for that. This will still have to work with pre/postZone etc... because those will only be cleaned up later.

**Comment by cbuescher** [comment]
@orenash I'm going to try to backport some of the changes I did in #9637 on 2.0 to the 1.x branch which will likely adress this issue here. So if you'd like to post a PR for your tests I could include these in the branches as run the backported changes against them as well.

**Comment by orenash** [comment]
Alright, thanks a lot! I will do so in the next days.

**Comment by cbuescher** [comment]
Hi @orenash, I included an integration tests based on your first comment in #9760, still would like to include your aditional rounding tests if you want to put them into a separate PR.

**Comment by orenash** [comment]
Don't you prefer to have the tests in the same PR with your fix? If you do not, I will open a PR later today.

**Comment by cbuescher** [comment]
No, this is great, will merge that after the fix is on the branch. Many thanks.

**Comment by cbuescher** [comment]
Fixed on 1.x with 78f0202 and on 1.4 with b7dbf1e
